### DISCUSSION TITLE
Need some help in analyzing a mixed traceback

### DISCUSSION THREAD

**Discussion by texadactyl** [ebitengine/purego#276] [Q&A] [answered]
Using github.com/ebitengine/purego v0.7.1.

The purpose of the attached Go code is to execute the JVM library libzip.so deflater from one byte array to another. This is a proof of concept.

Good news:
```
2024/08/31 12:33:34 Libraries jvm, zip, and java loaded ok
2024/08/31 12:33:34 purego.RegisterLibFunc (JNI_CreateJavaVM) ok
2024/08/31 12:33:34 createJvm ok
2024/08/31 12:33:34 purego.RegisterLibFunc (JNU_GetEnv) ok
2024/08/31 12:33:34 getEnv ok
2024/08/31 12:33:34 purego.RegisterLibFunc (Java_java_util_zip_Deflater_init) ok
2024/08/31 12:33:34 Java_java_util_zip_Deflater_init ok
2024/08/31 12:33:34 purego.RegisterLibFunc (Java_java_util_zip_Deflater_deflateBytesBytes) ok
```
So far, so good. I was able to load 3 required libraries and execute a few functions.
* Create a JVM.
* Get the pointer to the JNI environment.
* Initialize the deflater stream.

However, when I attempt to execute one function (the actual deflater), I see this:
```
SIGSEGV: segmentation violation
PC=0x717f52ba080d m=0 sigcode=1
signal arrived during cgo execution

goroutine 1 [syscall]:
runtime.cgocall(0x492620, 0xc000096410)
	/opt/go/src/runtime/cgocall.go:157 +0x4b fp=0xc0001133a0 sp=0xc000113368 pc=0x40596b
github.com/ebitengine/purego.RegisterFunc.func4({0xc0000b8000?, 0xb?, 0xb?})
	/home/elkins/go/pkg/mod/github.com/ebitengine/purego@v0.7.1/func.go:302 +0xc5f fp=0xc000113820 sp=0xc0001133a0 pc=0x48fd1f
reflect.callReflect(0xc000090210, 0xc000113d80, 0xc000113c58, 0xc000113c60)
	/opt/go/src/reflect/value.go:782 +0x54b fp=0xc000113c08 sp=0xc000113820 pc=0x473c4b
reflect.callReflect(0xc000090210, 0xc000113d80, 0xc000113c58, 0xc000113c60)
	<autogenerated>:1 +0x45 fp=0xc000113c38 sp=0xc000113c08 pc=0x4786e5
reflect.makeFuncStub()
	/opt/go/src/reflect/asm_amd64.s:47 +0x6e fp=0xc000113d80 sp=0xc000113c38 pc=0x47810e
main.main()
	/home/elkins/BASIS/jacomisc/misc.go/main.go:127 +0x8ef fp=0xc000113f40 sp=0xc000113d80 pc=0x49588f
runtime.main()
	/opt/go/src/runtime/proc.go:267 +0x2bb fp=0xc000113fe0 sp=0xc000113f40 pc=0x43561b
runtime.goexit()
	/opt/go/src/runtime/asm_amd64.s:1650 +0x1 fp=0xc000113fe8 sp=0xc000113fe0 pc=0x45f901

goroutine 2 [force gc (idle)]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
	/opt/go/src/runtime/proc.go:398 +0xce fp=0xc00004efa8 sp=0xc00004ef88 pc=0x435a6e
runtime.goparkunlock(...)
	/opt/go/src/runtime/proc.go:404
runtime.forcegchelper()
	/opt/go/src/runtime/proc.go:322 +0xb3 fp=0xc00004efe0 sp=0xc00004efa8 pc=0x4358f3
runtime.goexit()
	/opt/go/src/runtime/asm_amd64.s:1650 +0x1 fp=0xc00004efe8 sp=0xc00004efe0 pc=0x45f901
created by runtime.init.6 in goroutine 1
	/opt/go/src/runtime/proc.go:310 +0x1a

goroutine 3 [GC sweep wait]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
	/opt/go/src/runtime/proc.go:398 +0xce fp=0xc00004f778 sp=0xc00004f758 pc=0x435a6e
runtime.goparkunlock(...)
	/opt/go/src/runtime/proc.go:404
runtime.bgsweep(0x0?)
	/opt/go/src/runtime/mgcsweep.go:280 +0x94 fp=0xc00004f7c8 sp=0xc00004f778 pc=0x422414
runtime.gcenable.func1()
	/opt/go/src/runtime/mgc.go:200 +0x25 fp=0xc00004f7e0 sp=0xc00004f7c8 pc=0x4175c5
runtime.goexit()
	/opt/go/src/runtime/asm_amd64.s:1650 +0x1 fp=0xc00004f7e8 sp=0xc00004f7e0 pc=0x45f901
created by runtime.gcenable in goroutine 1
	/opt/go/src/runtime/mgc.go:200 +0x66

goroutine 4 [GC scavenge wait]:
runtime.gopark(0xc000028070?, 0x4d3fc0?, 0x1?, 0x0?, 0xc0000071e0?)
	/opt/go/src/runtime/proc.go:398 +0xce fp=0xc00004ff70 sp=0xc00004ff50 pc=0x435a6e
runtime.goparkunlock(...)
	/opt/go/src/runtime/proc.go:404
runtime.(*scavengerState).park(0x5520c0)
	/opt/go/src/runtime/mgcscavenge.go:425 +0x49 fp=0xc00004ffa0 sp=0xc00004ff70 pc=0x41fce9
runtime.bgscavenge(0x0?)
	/opt/go/src/runtime/mgcscavenge.go:653 +0x3c fp=0xc00004ffc8 sp=0xc00004ffa0 pc=0x42027c
runtime.gcenable.func2()
	/opt/go/src/runtime/mgc.go:201 +0x25 fp=0xc00004ffe0 sp=0xc00004ffc8 pc=0x417565
runtime.goexit()
	/opt/go/src/runtime/asm_amd64.s:1650 +0x1 fp=0xc00004ffe8 sp=0xc00004ffe0 pc=0x45f901
created by runtime.gcenable in goroutine 1
	/opt/go/src/runtime/mgc.go:201 +0xa5

goroutine 5 [finalizer wait]:
runtime.gopark(0x0?, 0x0?, 0x0?, 0x0?, 0x0?)
	/opt/go/src/runtime/proc.go:398 +0xce fp=0xc000050620 sp=0xc000050600 pc=0x435a6e
runtime.runfinq()
	/opt/go/src/runtime/mfinal.go:193 +0x107 fp=0xc0000507e0 sp=0xc000050620 pc=0x4165e7
runtime.goexit()
	/opt/go/src/runtime/asm_amd64.s:1650 +0x1 fp=0xc0000507e8 sp=0xc0000507e0 pc=0x45f901
created by runtime.createfing in goroutine 1
	/opt/go/src/runtime/mfinal.go:163 +0x3d

rax    0x10
rbx    0x1ee0e40
rcx    0x78
rdx    0x2
rdi    0x10
rsi    0xda78
rbp    0x1ee0dc0
rsp    0x7ffe6a811938
r8     0x0
r9     0x0
r10    0x717f52a11f10
r11    0x717f52ba0730
r12    0x2
r13    0x2
r14    0xc0000a0060
r15    0x1d67060
rip    0x717f52ba080d
rflags 0x10202
cs     0x33
fs     0x0
gs     0x0
exit status 2

```

I am not asking you to debug the attached program (main.go). I am looking for advice as to how one would diagnose what is going wrong in main.go or what am I supposed to gleam from the stack trace.

cc: @platypusguy
[main.go.zip](https://github.com/user-attachments/files/16825555/main.go.zip)

**Comment by texadactyl** [comment]
For whatever it is worth, I had desk-checked. Also, here are the variable values going into the ```deflateBytesToBytes``` call:
```
err = {error} nil
HandleJVM = {main.JVMAddr} 123558257738976
HandleENV = {main.JNIEnv} 18653280
dummyObject = {[]uint8} len:3, cap:3
dummyObjPtr = {main.jobject} 824634842608
libjvm = {uintptr} 18393120
libjava = {uintptr} 18560800
libzip = {uintptr} 18564992
funcName = {string} "Java_java_util_zip_Deflater_deflateBytesBytes"
createJvm = {func(*main.JVMAddr, *main.JNIEnv, *main.t_JavaVMInitArgs) main.jint} reflect.makeFuncStub
ret = {main.jint} 0
getEnv = {func(main.JVMAddr, *main.JNIEnv, main.jint) main.jint} reflect.makeFuncStub
deflaterInit = {func(main.JNIEnv, main.jobject, main.jint, main.jint, main.jboolean) main.jlong} reflect.makeFuncStub
level = {main.jint} 9
strategy = {main.jint} 0
noWrap = {main.jboolean} 0
streamAddr = {main.jlong} 20166128
deflateBytesToBytes = {func(main.JNIEnv, main.jobject, main.jlong, main.jbyteArray, main.jint, main.jint, main.jbyteArray, main.jint, main.jint, main.jint, main.jint) main.jlong} reflect.makeFuncStub
thisObject = {main.jobject} 824634940912
input = {[]uint8} len:26, cap:26
insize = {main.jint} 26
output = {[]uint8} len:26, cap:26
flush = {main.jint} 3
params = {main.jint} 0
inPtr = {main.jbyteArray} 824634940358
outPtr = {main.jbyteArray} 824633885024
```

**Comment by TotallyGamerJet** [comment]
First thanks for using purego.

The stack trace is almost always going to be useless in these situations since the issue is most likely caused by a disagreement of where the library expects an argument and where purego puts it. This usually happens because the function you are calling was defined incorrectly in the RegisterFunc function. 

I would try the following to solve this issue:
- try to the current alpha version of purego (maybe it's a bug we've fixed)
- Visually compare the C function you are trying to call and see if the arguments are the same size in Go code
- If you don't see a difference then write a C program that does what you want and use the sizeof operator for each argument to the failing function. And do the same in Go using unsafe.Sizeof and make they all match
- If that all fails you're left having to use a debugger and figure out why the arguments are different.

Also I'm not sure I understand why you are using the jvm to call into libzip when you can just download the C libzip.so from there website: https://libzip.org/ and run it directly without a jvm

**Reply by texadactyl** [reply]
Thank you for your reply and suggestions.

I agree that one should never use the Java library functions from Go to perform functions (E.g. zip). I am quite familiar with the POSIX libzip functions. And, a far better choice from Go would be to use one of the Go standard library functions such as compress/gzip.

My partner (@platypusguy) and I are engaged in the development of a Go-language JVM (https://github.com/platypusguy/jacobin). The source code that I attached here was the start of a proof of concept to see if we could use purego as a toolset for calling the C-language native functions. Given the volume, the eventual facility we adopt must be table-driven. Sounds like pure torture? (:

To date, most of the Jacobin compute work is handled in Go using a large switch structure in jvm/run.go, the key of which being the JVM operation code - see https://docs.oracle.com/javase/specs/jvms/se17/html/jvms-6.html for documetation of the JVM instruction set. This approach is working quite well, in general. We currently have a test base of 155 Java programs of various applications (https://github.com/texadactyl/jacotest.go) and most of the individual test cases pass.

Except for native functions (trapped and cause an abort in Jacobin), we are successful at mapping many of the Java class-method-signature functions in a generic wrapper function (jvm/gfunctionExec.go) back to to Go replacement functions using a few table-driven techniques, driven from the INVOKE* switch cases. To-date, this is how we have been avoiding some of the native functions. But, given the sheer volume of native functions (>1000) and their complex relationship to the JVM class code (many-to-many), we were hoping to avoid "boiling the ocean". I guess that we will be engaged in a sizable effort no matter what we do.

I have been also looking at using CGO too although it does not readily facilitate a table-driven approach. The same might be said using purego; time will tell. We might have to continue our original generic wrapper approach in replacing Java library class code to avoid the native functions that are called within those classes.

Thanks, again. All feedback is appreciated.

**Reply by TotallyGamerJet** [reply]
Ahh that's super cool. Yeah totally go for it sounds like a useful project

**Reply by texadactyl** [reply, marked as answer]
I wrote a C-language equivalent and blew up in that too, the same JVM library function call, namely ```Java_java_util_zip_Deflater_deflateBytesBytes```. All the simpler function calls work just fine in C and Go with purego. 

What's different? I am trying to deflate (compress, in normal English) a byte array. It turns out that byte arrays need help from some functions that are accessed via a vector stored in the primary JVM area that is represented with a pointer. The C calls look like this example: ```(*env)->SetByteArrayRegion(env, inputArray, 0, sizeof(input), (jbyte*)input);```. There are several functions of the same form that are required to set up the input and extract the output for array-oriented functions. Egads!

Needless to say, Go cannot support such a C-particuliar interface. I've also created that type of vectoring in C and Assembler.  It makes local mods on the fly to a function fairly easy. That is also how hardware interrupts are generally implemented. Good stuff! 

Anyways, I just wanted to tell you that y'all are innocent!  (:
